Using the titanic data, in your classification-exercises repository, create a notebook, decision_tree.ipynb where you will do the following:

In [33]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.impute import SimpleImputer
from sklearn.model_selection import train_test_split
import pandas as pd
import numpy as np
import pandas as pd
from acquire import get_titanic_data

# Load Titanic data into a pandas DataFrame
df_titanic = get_titanic_data()

# Display the first 5 rows of the DataFrame
df_titanic.head()

passenger_id  survived  pclass     sex   age  sibsp  parch     fare  \
0             0         0       3    male  22.0      1      0   7.2500   
1             1         1       1  female  38.0      1      0  71.2833   
2             2         1       3  female  26.0      0      0   7.9250   
3             3         1       1  female  35.0      1      0  53.1000   
4             4         0       3    male  35.0      0      0   8.0500   

  embarked  class deck  embark_town  alone  
0        S  Third  NaN  Southampton      0  
1        C  First    C    Cherbourg      0  
2        S  Third  NaN  Southampton      1  
3        S  First    C  Southampton      0  
4        S  Third  NaN  Southampton      1

In [2]:

# What is your baseline prediction? What is your baseline accuracy? remember: your baseline prediction for a classification problem is 
# predicting the most prevelant class in the training dataset (the mode). When you make those predictions, 
# what is your accuracy? This is your baseline accuracy.
# Count the number of passengers who did not survive
num_not_survived = df_titanic['survived'].value_counts()[0]

# Calculate the total number of passengers
total_passengers = len(df_titanic)

# Calculate the baseline accuracy
baseline_accuracy = num_not_survived / total_passengers

print("Baseline accuracy:", baseline_accuracy)


Baseline accuracy: 0.6161616161616161


In [39]:
# Fit the decision tree classifier to your training sample and transform (i.e. make predictions on the training sample)


# Load the titanic dataset
df_titanic = pd.read_csv('titanic.csv')

# Select the features and target variable
X = df_titanic.drop(columns=['survived'])
y = df_titanic['survived']

# One-hot encode categorical variables
X = pd.get_dummies(X)

# Split the dataset into training and testing subsets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Impute missing values with median
imp = SimpleImputer(strategy='median')
X_train = imp.fit_transform(X_train)
X_test = imp.transform(X_test)

# Initialize the decision tree classifier
dt = DecisionTreeClassifier(random_state=42)

# Fit the classifier to the training sample
dt.fit(X_train, y_train)

# Make predictions on the testing sample
y_pred_test = dt.predict(X_test)

# Calculate the accuracy of the classifier on the testing sample
accuracy_test = accuracy_score(y_test, y_pred_test)

print("Accuracy on testing sample:", accuracy_test)


Accuracy on testing sample: 0.7597765363128491


In [40]:
# Evaluate your in-sample results using the model score, confusion matrix, and classification report.

from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.impute import SimpleImputer
from sklearn.model_selection import train_test_split
import pandas as pd
import numpy as np

# Load the titanic dataset
df_titanic = pd.read_csv('titanic.csv')

# Select the features and target variable
X = df_titanic.drop(columns=['survived'])
y = df_titanic['survived']

# One-hot encode categorical variables
X = pd.get_dummies(X)

# Split the dataset into training and testing subsets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Impute missing values with median
imp = SimpleImputer(strategy='median')
X_train = imp.fit_transform(X_train)
X_test = imp.transform(X_test)

# Initialize the decision tree classifier
dt = DecisionTreeClassifier(random_state=42)

# Fit the classifier to the training sample
dt.fit(X_train, y_train)

# Make predictions on the training and testing samples
y_pred_train = dt.predict(X_train)
y_pred_test = dt.predict(X_test)

# Evaluate the accuracy of the classifier on the training and testing samples
accuracy_train = dt.score(X_train, y_train)
accuracy_test = dt.score(X_test, y_test)

print("Accuracy on training sample:", accuracy_train)
print("Accuracy on testing sample:", accuracy_test)

# Evaluate the confusion matrix and classification report on the testing sample
print("Confusion matrix on testing sample:\n", confusion_matrix(y_test, y_pred_test))
print("Classification report on testing sample:\n", classification_report(y_test, y_pred_test))


# The `score` method returns the mean accuracy on the given data and labels, while the `confusion_matrix` function computes the confusion matrix to evaluate the accuracy of a classification. The `classification_report` function computes several classification metrics, including precision, recall, f1-score, and support, for each class.

# The results show that the model achieves perfect accuracy on the training sample, but only 78.8% accuracy on the testing sample,
# which indicates some overfitting to the training data. The confusion matrix and classification report show that the model performs 
# reasonably well on both classes, but has a slightly higher precision and recall for the negative class (did not survive) compared 
# to the positive class (survive Overall, the model can benefit from some parameter tuning to improve its performance on the testing data.

Accuracy on training sample: 1.0
Accuracy on testing sample: 0.7597765363128491
Confusion matrix on testing sample:
 [[82 23]
 [20 54]]
Classification report on testing sample:
               precision    recall  f1-score   support

           0       0.80      0.78      0.79       105
           1       0.70      0.73      0.72        74

    accuracy                           0.76       179
   macro avg       0.75      0.76      0.75       179
weighted avg       0.76      0.76      0.76       179



In [41]:
# Compute: Accuracy, true positive rate, false positive rate, true negative rate, false negative rate, precision, recall, f1-score, and support.

from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.impute import SimpleImputer
from sklearn.model_selection import train_test_split
import pandas as pd
import numpy as np

# Load the titanic dataset
df_titanic = pd.read_csv('titanic.csv')

# Select the features and target variable
X = df_titanic.drop(columns=['survived'])
y = df_titanic['survived']

# One-hot encode categorical variables
X = pd.get_dummies(X)

# Split the dataset into training and testing subsets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Impute missing values with median
imp = SimpleImputer(strategy='median')
X_train = imp.fit_transform(X_train)
X_test = imp.transform(X_test)

# Initialize the decision tree classifier
dt = DecisionTreeClassifier(random_state=42)

# Fit the classifier to the training sample
dt.fit(X_train, y_train)

# Make predictions on the testing sample
y_pred_test = dt.predict(X_test)

# Compute the accuracy, confusion matrix, and classification report on the testing sample
accuracy = accuracy_score(y_test, y_pred_test)
tn, fp, fn, tp = confusion_matrix(y_test, y_pred_test).ravel()
tpr = tp / (tp + fn)
fpr = fp / (fp + tn)
tnr = tn / (tn + fp)
fnr = fn / (fn + tp)
precision = tp / (tp + fp)
recall = tpr
f1_score = 2 * (precision * recall) / (precision + recall)
support = np.bincount(y_test)

print("Accuracy:", accuracy)
print("True positive rate (TPR):", tpr)
print("False positive rate (FPR):", fpr)
print("True negative rate (TNR):", tnr)
print("False negative rate (FNR):", fnr)
print("Precision:", precision)
print("Recall:", recall)
print("F1-score:", f1_score)
print("Support:", support)


Accuracy: 0.7597765363128491
True positive rate (TPR): 0.7297297297297297
False positive rate (FPR): 0.21904761904761905
True negative rate (TNR): 0.780952380952381
False negative rate (FNR): 0.2702702702702703
Precision: 0.7012987012987013
Recall: 0.7297297297297297
F1-score: 0.7152317880794701
Support: [105  74]


In [42]:
# Run through steps 2-4 using a different max_depth value.


from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.impute import SimpleImputer
from sklearn.model_selection import train_test_split
import pandas as pd
import numpy as np

# Load the titanic dataset
df_titanic = pd.read_csv('titanic.csv')

# Select the features and target variable
X = df_titanic.drop(columns=['survived'])
y = df_titanic['survived']

# One-hot encode categorical variables
X = pd.get_dummies(X)

# Split the dataset into training and testing subsets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Impute missing values with median
imp = SimpleImputer(strategy='median')
X_train = imp.fit_transform(X_train)
X_test = imp.transform(X_test)

# Initialize the decision tree classifier with max_depth=3
dt_depth3 = DecisionTreeClassifier(max_depth=3, random_state=42)

# Fit the classifier to the training sample
dt_depth3.fit(X_train, y_train)

# Make predictions on the testing sample
y_pred_test = dt_depth3.predict(X_test)

# Compute the accuracy, confusion matrix, and classification report on the testing sample
accuracy = accuracy_score(y_test, y_pred_test)
tn, fp, fn, tp = confusion_matrix(y_test, y_pred_test).ravel()
tpr = tp / (tp + fn)
fpr = fp / (fp + tn)
tnr = tn / (tn + fp)
fnr = fn / (fn + tp)
precision = tp / (tp + fp)
recall = tpr
f1_score = 2 * (precision * recall) / (precision + recall)
support = np.bincount(y_test)

print("Accuracy:", accuracy)
print("True positive rate (TPR):", tpr)
print("False positive rate (FPR):", fpr)
print("True negative rate (TNR):", tnr)
print("False negative rate (FNR):", fnr)
print("Precision:", precision)
print("Recall:", recall)
print("F1-score:", f1_score)
print("Support:", support)


Accuracy: 0.7988826815642458
True positive rate (TPR): 0.6891891891891891
False positive rate (FPR): 0.12380952380952381
True negative rate (TNR): 0.8761904761904762
False negative rate (FNR): 0.3108108108108108
Precision: 0.796875
Recall: 0.6891891891891891
F1-score: 0.7391304347826088
Support: [105  74]


In [43]:
# Which model performs better on your in-sample data?
# Compute the accuracy of the default model on the training sample
y_pred_train_default = dt.predict(X_train)
accuracy_default = accuracy_score(y_train, y_pred_train_default)

# Compute the accuracy of the model with max_depth=3 on the training sample
y_pred_train_depth3 = dt_depth3.predict(X_train)
accuracy_depth3 = accuracy_score(y_train, y_pred_train_depth3)

print("Accuracy of default model on training sample:", accuracy_default)
print("Accuracy of model with max_depth=3 on training sample:", accuracy_depth3)


# If the output shows that both the default model and the model with `max_depth=3` have the same accuracy of 83.4% on the training sample, 
# it suggests that the model with `max_depth=3` is not overfitting to the training data as much as the default model. 
# This may be due to the fact that the `max_depth` hyperparameter is limiting the complexity of the model and preventing it 
# from memorizing the training data too well.

# In this case, we may prefer the model with `max_depth=3` over the default model because it is likely to generalize better to new,
#  unseen data. However, we should still evaluate the models on the testing sample to confirm their performance and choose the one with
#  the best overall performance.

Accuracy of default model on training sample: 1.0
Accuracy of model with max_depth=3 on training sample: 0.8342696629213483


In [44]:
def split_data(df, target):
    '''
    take in a DataFrame and return train, validate, and test DataFrames; stratify on target variable.
    return train, validate, test DataFrames.
    '''
    train_validate, test = train_test_split(df, test_size=.2, random_state=123, stratify=df[target])
    train, validate = train_test_split(train_validate, 
                                       test_size=.3, 
                                       random_state=123, 
                                       stratify=train_validate[target])
    return train, validate, test
import pandas as pd
from acquire import get_titanic_data
from sklearn.model_selection import train_test_split

target = 'survived'
train_titanic, validate_titanic, test_titanic = split_data(df_titanic, target)

# Print the three datasets

print(f'Train: {train_titanic.shape}, Validate: {validate_titanic.shape}, Test: {test_titanic.shape}')

Train: (498, 14), Validate: (214, 14), Test: (179, 14)


### Which model performs best on your out-of-sample data, the validate set?

Based on the test set accuracies, the model with `max_depth=3` performs better on the out-of-sample data than the default model. The accuracy of the model with `max_depth=3` is 0.7989 while the accuracy of the default model is 0.7598. Therefore, the model with `max_depth=3` is likely the better choice for this task.

In [45]:
# Work through these same exercises using the Telco dataset.

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from acquire import get_telco_data

# Load Telco data into a pandas DataFrame
df_telco = get_telco_data()
df_telco = df_telco.drop(['payment_type_id', 'internet_service_type_id', 'contract_type_id', 'customer_id'], axis=1)
df_telco.head()

gender  senior_citizen partner dependents  tenure phone_service  \
0  Female               0     Yes        Yes       9           Yes   
1    Male               0      No         No       9           Yes   
2    Male               0      No         No       4           Yes   
3    Male               1     Yes         No      13           Yes   
4  Female               1     Yes         No       3           Yes   

  multiple_lines online_security online_backup device_protection tech_support  \
0             No              No           Yes                No          Yes   
1            Yes              No            No                No           No   
2             No              No            No               Yes           No   
3             No              No           Yes               Yes           No   
4             No              No            No                No          Yes   

  streaming_tv streaming_movies paperless_billing  monthly_charges  \
0          Yes               No               Yes             65.6   
1           No              Yes                No             59.9   
2           No               No               Yes             73.9   
3          Yes              Yes               Yes             98.0   
4          Yes               No               Yes             83.9   

  total_charges churn   contract_type internet_service_type      payment_type  
0         593.3    No        One year                   DSL      Mailed check  
1         542.4    No  Month-to-month                   DSL      Mailed check  
2        280.85   Yes  Month-to-month           Fiber optic  Electronic check  
3       1237.85   Yes  Month-to-month           Fiber optic  Electronic check  
4         267.4   Yes  Month-to-month           Fiber optic      Mailed check

In [46]:

# Determine the most prevalent class in the 'churn' column
mode = df_telco['churn'].mode()[0]

# Make all predictions using the most prevalent class
baseline_preds = np.full(df_telco.shape[0], mode)

# Calculate the accuracy of the baseline model
baseline_accuracy = (baseline_preds == df_telco['churn']).mean()

print(f"Baseline Prediction: {mode}")
print(f"Baseline Accuracy: {baseline_accuracy:.4f}")


Baseline Prediction: No
Baseline Accuracy: 0.7346


In [47]:


from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.impute import SimpleImputer
from sklearn.model_selection import train_test_split
import pandas as pd
import numpy as np

# Select the features and target variable
X = df_telco.drop(columns=['churn'])
y = df_telco['churn']

# One-hot encode categorical variables
X = pd.get_dummies(X)

# Split the dataset into training and testing subsets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Impute missing values with median
imp = SimpleImputer(strategy='median')
X_train = imp.fit_transform(X_train)
X_test = imp.transform(X_test)

# Initialize the decision tree classifier
dt = DecisionTreeClassifier(random_state=42)

# Fit the classifier to the training sample
dt.fit(X_train, y_train)

# Make predictions on the training and testing samples
y_pred_train = dt.predict(X_train)
y_pred_test = dt.predict(X_test)

# Evaluate the accuracy of the classifier on the training and testing samples
accuracy_train = dt.score(X_train, y_train)
accuracy_test = dt.score(X_test, y_test)

print("Accuracy on training sample:", accuracy_train)
print("Accuracy on testing sample:", accuracy_test)

# Evaluate the confusion matrix and classification report on the testing sample
print("Confusion matrix on testing sample:\n", confusion_matrix(y_test, y_pred_test))
print("Classification report on testing sample:\n", classification_report(y_test, y_pred_test)) 


Accuracy on training sample: 0.9978700745473909
Accuracy on testing sample: 0.758694109297374
Confusion matrix on testing sample:
 [[893 143]
 [197 176]]
Classification report on testing sample:
               precision    recall  f1-score   support

          No       0.82      0.86      0.84      1036
         Yes       0.55      0.47      0.51       373

    accuracy                           0.76      1409
   macro avg       0.69      0.67      0.67      1409
weighted avg       0.75      0.76      0.75      1409



In [29]:

from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.impute import SimpleImputer
from sklearn.model_selection import train_test_split
import pandas as pd
import numpy as np


# Convert target variable to numeric values
df_telco['churn'] = pd.factorize(df_telco['churn'])[0]

# Select the features and target variable
X = df_telco.drop(columns=['churn'])
y = df_telco['churn']

# One-hot encode categorical variables
X = pd.get_dummies(X)

# Split the dataset into training, validation, and testing subsets
X_train_val, X_test, y_train_val, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
X_train, X_val, y_train, y_val = train_test_split(X_train_val, y_train_val, test_size=0.25, random_state=42)

# Impute missing values with median
imp = SimpleImputer(strategy='median')
X_train = imp.fit_transform(X_train)
X_val = imp.transform(X_val)
X_test = imp.transform(X_test)

# Initialize the decision tree classifier
dt = DecisionTreeClassifier(random_state=42)

# Fit the classifier to the training sample
dt.fit(X_train, y_train)

# Make predictions on the validation sample
y_pred_val = dt.predict(X_val)

# Compute the accuracy, confusion matrix, and classification report on the validation sample
accuracy = accuracy_score(y_val, y_pred_val)
tn, fp, fn, tp = confusion_matrix(y_val, y_pred_val).ravel()
tpr = tp / (tp + fn)
fpr = fp / (fp + tn)
tnr = tn / (tn + fp)
fnr = fn / (fn + tp)
precision = tp / (tp + fp)
recall = tpr
f1_score = 2 * (precision * recall) / (precision + recall)
support = np.bincount(y_val)

print("Accuracy on validation sample:", accuracy)
print("True positive rate (TPR):", tpr)
print("False positive rate (FPR):", fpr)
print("True negative rate (TNR):", tnr)
print("False negative rate (FNR):", fnr)
print("Precision:", precision)
print("Recall:", recall)
print("F1-score:", f1_score)
print("Support:", support)

# Make predictions on the testing sample
y_pred_test = dt.predict(X_test)

# Compute the accuracy, confusion matrix, and classification report on the testing sample
accuracy = accuracy_score(y_test, y_pred_test)
tn, fp, fn, tp = confusion_matrix(y_test, y_pred_test).ravel()
tpr = tp / (tp + fn)
fpr = fp / (fp + tn)
tnr = tn / (tn + fp)
fnr = fn / (fn + tp)
precision = tp / (tp + fp)
recall = tpr
f1_score = 2 * (precision * recall) / (precision + recall)
support = np.bincount(y_test)

print("Accuracy on testing sample:", accuracy)
print("True positive rate (TPR):", tpr)
print("False positive rate (FPR):", fpr)
print("True negative rate (TNR):", tnr)
print("False negative rate (FNR):", fnr)
print("Precision:", precision)
print("Recall:", recall)
print("F1-score:", f1_score)
print("Support:", support)



Accuracy on validation sample: 0.7693399574166075
True positive rate (TPR): 0.48320413436692505
False positive rate (FPR): 0.1223091976516634
True negative rate (TNR): 0.8776908023483366
False negative rate (FNR): 0.5167958656330749
Precision: 0.5993589743589743
Recall: 0.48320413436692505
F1-score: 0.5350500715307582
Support: [1022  387]
Accuracy on testing sample: 0.7544357700496807
True positive rate (TPR): 0.47989276139410186
False positive rate (FPR): 0.14671814671814673
True negative rate (TNR): 0.8532818532818532
False negative rate (FNR): 0.5201072386058981
Precision: 0.540785498489426
Recall: 0.47989276139410186
F1-score: 0.5085227272727274
Support: [1036  373]


In [31]:
# Run through steps 2-4 using a different max_depth value.

from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.impute import SimpleImputer
from sklearn.model_selection import train_test_split
import pandas as pd
import numpy as np


# Convert target variable to numeric values
df_telco['churn'] = pd.factorize(df_telco['churn'])[0]

# Select the features and target variable
X = df_telco.drop(columns=['churn'])
y = df_telco['churn']

# One-hot encode categorical variables
X = pd.get_dummies(X)

# Split the dataset into training and testing subsets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Impute missing values with median
imp = SimpleImputer(strategy='median')
X_train = imp.fit_transform(X_train)
X_test = imp.transform(X_test)

# Initialize the decision tree classifier with max_depth=3
dt_depth3 = DecisionTreeClassifier(max_depth=3, random_state=42)

# Fit the classifier to the training sample
dt_depth3.fit(X_train, y_train)

# Make predictions on the testing sample
y_pred_test = dt_depth3.predict(X_test)

# Compute the accuracy, confusion matrix, and classification report on the testing sample
accuracy = accuracy_score(y_test, y_pred_test)
tn, fp, fn, tp = confusion_matrix(y_test, y_pred_test).ravel()
tpr = tp / (tp + fn)
fpr = fp / (fp + tn)
tnr = tn / (tn + fp)
fnr = fn / (fn + tp)
precision = tp / (tp + fp)
recall = tpr
f1_score = 2 * (precision * recall) / (precision + recall)
support = np.bincount(y_test)

print("Accuracy:", accuracy)
print("True positive rate (TPR):", tpr)
print("False positive rate (FPR):", fpr)
print("True negative rate (TNR):", tnr)
print("False negative rate (FNR):", fnr)
print("Precision:", precision)
print("Recall:", recall)
print("F1-score:", f1_score)
print("Support:", support)


Accuracy: 0.7835344215755855
True positive rate (TPR): 0.3351206434316354
False positive rate (FPR): 0.05501930501930502
True negative rate (TNR): 0.944980694980695
False negative rate (FNR): 0.6648793565683646
Precision: 0.6868131868131868
Recall: 0.3351206434316354
F1-score: 0.45045045045045046
Support: [1036  373]


In [32]:

from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.impute import SimpleImputer
import pandas as pd

# Impute missing values with median
imp = SimpleImputer(strategy='median')
X = imp.fit_transform(X)

# Initialize the decision tree classifier with default parameters
dt = DecisionTreeClassifier(random_state=42)

# Fit the classifier to the data
dt.fit(X, y)

# Initialize the decision tree classifier with max_depth=3
dt_depth3 = DecisionTreeClassifier(max_depth=3, random_state=42)

# Fit the classifier to the data
dt_depth3.fit(X, y)

# Make predictions on the training sample
y_pred_train_default = dt.predict(X)
accuracy_default = accuracy_score(y, y_pred_train_default)

# Compute the accuracy of the model with max_depth=3 on the training sample
y_pred_train_depth3 = dt_depth3.predict(X)
accuracy_depth3 = accuracy_score(y, y_pred_train_depth3)

print("Accuracy of default model on training sample:", accuracy_default)
print("Accuracy of model with max_depth=3 on training sample:", accuracy_depth3)

#model with the max depth of 3 is the better model so it is not overfitting.
# this model performs the best with all the data samples

Accuracy of default model on training sample: 1.0
Accuracy of model with max_depth=3 on training sample: 0.7911401391452506


Experiment with this model on other datasets with a higher number of output classes.(Bonus per Misty)